Automattic/jetpack
# Publicize: Ensure that connections are only retrieved once per request.

A race condition where the transient, used to cache the connections information, gets set in a second request, while another is in flight, causes the `set_transient` call to fail.

As we relied on the transient, this meant that each time we tried to retrieve the current connection information, we would make a request to the API, and then fail to set the transient again.

This change solves that problem by storing the information in memory and also removing the transient if the call to `set_transient` fails.

There is an outstanding task where we need to work out why the Connections_Post_Field causes so many calls to `get_all_connections`.

### Other information:


## Jetpack product discussion
p1719382901926009-slack-C02JJ910CNL

## Does this pull request change what data or activity we track or use?


No.

## Testing instructions:





This is a bit tricky as it's a race condition, but it seems to be quite possible with a local dev site

* Make sure you have at least one social connection on the site
* Create a test post, and line up your cursor with the left-hand side of the publish button (I know!)
* Click the button, and then quickly click the second Publish button that appears
* Without this change, the request will take quite a bit of time, and the post will not be shared to social media
* With this change, the requests should complete quickly, and the post will be shared as usual.

A video of this is in the linked Slack thread.
